doc/design/async_devices.md

```diff
@@ -0,0 +1,68 @@
+# Design Doc: Asynchronous and Non-blocking techniques on heterogeneous devices.
+
+
+## Problem
+
+We often use heterogeneous devices, like GPU, FPGA, in a deep learning platform. The computation on heterogeneous devices is usually asynchronous and non-blocking. The computation tasks are offloaded to heterogeneous devices and programmers must wait for the computation finished and then fetch results. Heterogeneous devices are usually able to execute simultaneously. For example, Fermi architecture (CUDA compute capability 2.0+) can simultaneously support
+
+* execute CUDA kernels
+* host-to-device memory copy
+* device-to-host memory copy.
+
+However, the simultaneous execution is not transparent to programmers. 
+
+Let's use CUDA as an example. There is a building block named `stream` in CUDA. Streams introduce task-based parallelism to CUDA codes. The sequence of operations will be executed in issue-order on the GPU if they are in the same stream. 
+
+The operators in different streams are able to run concurrently as long as they are in multiple streams and hardware supports it. CUDA hardware has no notion of streams. The hardware has separate queues (engines) to perform memory copies and to execute kernels.
+
+If we want to take advantage of CUDA devices, we must use at least N streams, where N equals the number of hardware queues, and separate operators into these streams. The N equals to three since CUDA can simultaneously execute CUDA kernels, H2D memcpy, D2H memcpy by the CUDA hardware.
+
+Considering the execution of CUDA devices is asynchronous, there should be a wait operator when switching streams. For example, we want to read the computation result from GPU; then we must wait for the computation complete and issue a device-to-host memory copy.
+
+## Solution
+
+The solution is straightforward based on the hardware properties we described in the problem section. We should:
+
+* Create N device contexts on one device. The N should be corresponding to the hardware property. For example, the CUDA devices should have three device contexts.
+
+* Every tensor should hold the one device context, where the current operator of the tensor is performed on.
+
+* Wait for the execution complete on the previous device context, when switching the current device context of tensors.
+
+
+The sample C++ program is
+
+```cpp
+
+enum CUDAHardwareStream {
+  kCOMPUTATION,
+  kD2HMEMCPY,
+  kH2DMEMCPY
+};
+
+std::map<CUDAHardwareStream, DeviceContext* > gDevCtxs;
+
+class Tensor {
+public:
+  ...
+  
+  void SwitchDevCtx(DeviceContext* new_ctx) {
```

Comment:
I am hesitant to add `SwitchDevCtx` as a method of Tensor. Reasons:

1. If we have`Tensor::SwitchDevCtx`, we may also need `SelectedRows::SwitchDevCtx` etc.
1. The operator needs to wait, not the tensor.

So maybe we should put the explicit wait in operator run?

```c++
void ReduceOp::Run(scope, place) {
    gDevCtxs[place, kCOMPUTATION].Wait();
    my_ctx = gDevCtx[place, kD2DMEMCPY];

    ...   
}
```

Reply:
Only data in different CUDA streams are independent, can these operations execute `potentially` in parallel. So, the basic problem is to analyze data dependencies between two operations. Maybe we need an explicit scheduler module to do these.

Reply:
@tonyyang-svail 
I don't think `the operator needs to wait, not the tensor`.

The previous operator of the tensor needs to wait for. So we need

1. Record the previous operator of the tensor.
   * For example, `ReduceOp::Run` may not need to wait `kCOMPUTATION` if the previous operators are not computation. `ReduceOp::Run` could wait for any kind of device contexts.
   * The input tensors of `ReduceOp` can be operated by various streams. For example, there are two tensors need to be reduced, A, B. A is a computational result. B is a H2DMemcpy result. The two device contexts should be both waited.
2. Not all operators of Tensor performed by `paddle::framework::Operator`.
   * There are memcpy, fill zero in framework module.
   * We cannot just put an explicit wait in operator run to solve this problem

Reply:
> Only data in different CUDA streams are independent, can these operations execute potentially in parallel. So, the basic problem is to analyze data dependencies between two operations. Maybe we need an explicit scheduler module to do these.

@QiJune 
An explicit scheduler module could resolve these problems, however,

1. Not all operators of Tensor performed by `paddle::framework::Operator` and `Executor`.
   * A scheduler should schedule operators. However, there is no unified abstraction of operators on Tensor to be scheduled.
   * If we want to add a scheduler, we should add an abstraction layer of operators first.
2. There is no clear schedule algorithm for Fluid
   * Fluid is different from other frameworks. We do not use DAG to represent neural networks. Tensors and variables can be overwritten. There could be loops in our framework. We should give a clear schedule algorithm before we write it.
   * An explicit scheduler may NOT be faster than switching streams as it needs.
      * Switching streams as it needs will have more conditions in C++ (`if statements`). However, comparing than the computation and wait for streams, the conditions are ignorable.
      * A explicit scheduler also needs to calculate dependencies ahead of time. It is not free.

Reply:
@reyoung Maybe we can write some experimental codes. Following is a pseudocode：

```
tensor1 = op1(dev_ctx1);
tensor2 = op2(tensor1, dev_ctx1);
tensor3 = op3(tensor2, dev_ctx1);
tensor4 = update_op(tensor1, dev_ctx2);
```

We expect that after op1 running, update_op can run in parallel with op2 and op3. But `cudaStreamSynchronize` will block until stream has completed all operations. Please refer to  the [official doc](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g82b5784f674c17c6df64affe618bf45e). 

I am not sure if the behavior will be update_op running after op1/op2/op3 finishing, because the dev_ctx1 stream has three operations on it. That's not we wanted.

Reply:
@QiJune thanks for the example. I am sure these four operators will be executed sequentially.

As fas as parallel_do_grad is concerned, I think the following program is good enough, even without an explicit scheduler
```
parallel_do_grad
    w1_grad = fc_grad(.., stream0)
    all_reduce(w1_grad, stream1)    // it will wait for stream0
    sgd(w1_grad, w1, stream1)
    w2_grad = fc_grad(.., stream0)
    ...
```